File: SQCQPpack/SNSQP/demonRecovery.m
% demon recovery problems
clc; clear all; close all; 
addpath(genpath(pwd));
 
n          = 1000; 
k          = ceil(0.01*n); % number of quadratic const
m          = ceil(0.01*n); % number of linear const
s          = ceil(0.05*n); % sparsity level

test       = 1; % X = R^n        if test =1 
                % X = [-2,2]^n   if test =2 
                % X = [0,inf)^n  if test =3
switch test
  case 1  
       lb   = -inf;
       ub   = inf;         
       xT   = randn(s,1);
  case 2  
       lb   = -2;
       ub   = 2;
       xT   = unifrnd(lb,ub,[s,1]);
  case 3  
       lb   = 0;
       ub   = inf; 
       xT   = 1*rand(s,1); 
end 

T             = randperm(n,s); 
xopt          = zeros(n,1); 
xopt(T)       = xT;
dt            = DataRecovery(n,k,m,xopt,T); 

pars.x0       = zeros(n,1);
pars.tau      = 3; % decrease this value if the algorithm do not converge
pars.dualquad = 0.001*ones(k,1);
pars.dualineq = 0.001*ones(m,1);
pars.itlser   = 1; % increase this value if the algorithm varies a lot
out           = SNSQP(n,s,dt.Q0,dt.q0,dt.Qi,dt.qi,dt.ci,dt.A,dt.b,[],[],lb,ub,pars);
fprintf(' Relerr:    %7.3e \n', norm(out.sol-xopt)/norm(xopt)); 
PlotRecovery(xopt,out.sol,[900,500,500,250],1)

File: SQCQPpack/SNSQP/demonRecovery.py
import sys
from pathlib import Path

import numpy as np

_HERE = Path(__file__).resolve().parent
_PARENT = _HERE.parent
for candidate in (_PARENT, _HERE):
    if str(candidate) not in sys.path:
        sys.path.insert(0, str(candidate))

from SNSQP.solver import snsqp
from SNSQP.solver.PlotRecovery import plot_recovery
from SNSQP.solver.data.DataRecovery import generate_recovery_data


def main():
    rng = np.random.RandomState()

    n = 1000
    k = int(np.ceil(0.01 * n))
    m = int(np.ceil(0.01 * n))
    s = int(np.ceil(0.05 * n))

    test = 1
    if test == 1:
        lb = -np.inf
        ub = np.inf
        xT = rng.randn(s)
    elif test == 2:
        lb = -2.0
        ub = 2.0
        xT = rng.uniform(lb, ub, size=s)
    else:
        lb = 0.0
        ub = np.inf
        xT = rng.rand(s)

    support = rng.permutation(n)[:s]
    xopt = np.zeros(n)
    xopt[support] = xT

    data = generate_recovery_data(n, k, m, xopt, support, rng)

    pars = {
        'x0': np.zeros(n),
        'tau': 3.0,
        'dualquad': np.full(k, 0.001),
        'dualineq': np.full(m, 0.001),
        'itlser': 1,
    }

    out = snsqp(
        n,
        s,
        data['Q0'],
        data['q0'],
        data['Qi'],
        data['qi'],
        data['ci'],
        data['A'],
        data['b'],
        None,
        None,
        lb,
        ub,
        pars,
    )

    relerr = np.linalg.norm(out['sol'] - xopt) / (np.linalg.norm(xopt) + 1e-12)
    print(f" Relerr:    {relerr:7.3e} ")

    try:
        plot_recovery(xopt, out['sol'])
    except RuntimeError:
        pass


if __name__ == '__main__':
    main()
